raft/raft.go

```diff
@@ -0,0 +1,292 @@
+package raft
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/adylanrff/raft-algorithm/raft/model"
+	"github.com/adylanrff/raft-algorithm/rpc"
+	log "github.com/sirupsen/logrus"
+)
+
+type Raft interface {
+	RequestVote(req *model.RequestVoteRequestDTO) (*model.RequestVoteResponseDTO, error)
+	AppendEntries(req *model.AppendEntriesRequestDTO) (*model.AppendEntriesResponseDTO, error)
+}
+
+type defaultRaft struct {
+	address string
+	state   model.RaftState
+	config  RaftConfig
+
+	// timer
+	electionTimer   *time.Timer
+	heartbeatTicker *time.Ticker
+
+	raftClient RaftClient
+
+	appendEntriesSignal chan struct{}
+}
+
+func NewRaft(id string, address string, cfg RaftConfig) *defaultRaft {
+	electionTimeout := rand.Intn(int(cfg.ElectionTimeout*2)-int(cfg.ElectionTimeout)) + int(cfg.ElectionTimeout)
+	cfg.ElectionTimeout = time.Duration(electionTimeout)
+
+	fmt.Println(electionTimeout)
+	return &defaultRaft{
+		address:       address,
+		state:         model.NewRaftState(id),
+		config:        cfg,
+		electionTimer: time.NewTimer(cfg.ElectionTimeout),
+		raftClient:    NewRaftClient(rpc.NewRPCClient()),
+
+		appendEntriesSignal: make(chan struct{}),
+	}
+}
+
+// RequestVote implements RaftServerHandler
+func (r *defaultRaft) RequestVote(req *model.RequestVoteRequestDTO) (*model.RequestVoteResponseDTO, error) {
+	r.state.Lock()
+	defer r.state.Unlock()
+
+	log.WithFields(log.Fields{
+		"id":       r.state.ID,
+		"votedFor": r.state.VotedFor,
+		"req":      req.RequestVoteRequest,
+		"method":   "RequestVote",
+	}).Debug("request_vote_request_start")
+
+	// reply false if the term is less than our current term
+	if req.GetTerm() < r.state.CurrentTerm {
+		log.WithFields(log.Fields{
+			"candidate_term": req.GetTerm(),
+			"current_term":   r.state.CurrentTerm,
+		}).Error("outdated term")
+		return model.NewRequestVoteResponseDTO(r.state.CurrentTerm, false), nil
+	}
+
+	// Reply false also if we already voted and it is not the candidate ID
+	if r.state.VotedFor != req.GetCandidateId() && r.state.VotedFor != "" {
+		log.WithFields(log.Fields{
+			"voted_for":    r.state.VotedFor,
+			"candidate_id": req.GetCandidateId(),
+		}).Error("already voted before")
+		return model.NewRequestVoteResponseDTO(r.state.CurrentTerm, false), nil
+	}
+
```

Comment:
ada election restriction yang masih kurang mestinya disini. mestinya servernya juga ensure log mana yang lebih up-to-date given the last log index and term. but, given that we haven't done anything with the log yet, i think we can add it later.

Reply:
Oh yes bener, kemarin skip dulu karena lognya belum ada isinya haha